Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs

```diff
@@ -0,0 +1,238 @@
+            this.npmHandler = new NpmCommandHandler(this.outputPane);
+        }
+
+        public IChildrenSource ProvideChildren(WorkspaceVisualNodeBase parentNode)
+        {
+            // We don't provide new children so return null
+            return null;
+        }
+
+        public IWorkspaceCommandHandler ProvideCommandHandler(WorkspaceVisualNodeBase parentNode)
+        {
+            if (NpmCommandHandler.EnsurePackageJson(parentNode))
+            {
+                this.outputPane.ShowWindow();
+                return this.npmHandler;
+            }
+
+            return null;
+        }
+
+        private sealed class NpmCommandHandler : IWorkspaceCommandHandler
+        {
+            private readonly OutputPaneWrapper outputPane;
+
+            public NpmCommandHandler(OutputPaneWrapper outputPane)
+            {
+                this.outputPane = outputPane;
+            }
+
+            public int Priority => 1000;
+
+            public bool IgnoreOnMultiselect => true;
+
+            public int Exec(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
+            {
+                var node = selection.FirstOrDefault();
+                if (selection.Count != 1 || !EnsurePackageJson(node))
+                {
+                    return (int)Constants.OLECMDERR_E_NOTSUPPORTED;
+                }
+
+                if (pguidCmdGroup == Guids.NodeToolsWorkspaceCmdSet)
+                {
+                    switch (nCmdID)
+                    {
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallMissing:
+                            ExecNpmInstallMissing(node);
+                            return VSConstants.S_OK;
+
+                        case PkgCmdId.cmdidWorkSpaceNpmInstallNew:
+                            ExecNpmInstallNew(node);
+                            return VSConstants.S_OK;
+
+                        case PkgCmdId.cmdidWorkSpaceNpmUpdate:
+                            ExecNpmUpdate(node);
+                            return VSConstants.S_OK;
+                    }
+
+                    if (nCmdID >= PkgCmdId.cmdidWorkSpaceNpmDynamicScript && nCmdID < PkgCmdId.cmdidWorkSpaceNpmDynamicScriptMax)
+                    {
+                        ExecDynammic(node, nCmdID);
+                        return VSConstants.S_OK;
+                    }
+                }
+
+                return (int)Constants.OLECMDERR_E_NOTSUPPORTED;
+            }
+
+            // Note: all the Exec commands are async, this allows us to call them in a fire and forget
+            // pattern, without blocking the UI or losing any logging
+
+            private async void ExecNpmInstallMissing(WorkspaceVisualNodeBase node)
+            {
+                using (var npmController = this.CreateController(node.Workspace))
+                using (var commander = npmController.CreateNpmCommander())
+                {
+                    await commander.Install();
```

Comment:
Does this update the status bar or something?  How does the user know it's working?

Reply:
Yes, and write to the output window